chips/litex/src/uart.rs

```diff
@@ -0,0 +1,538 @@
+    /// LiteX EventManager pending register
+    ev_pending: R::ReadWrite8,
+    /// LiteX EventManager pending register
+    ev_enable: R::ReadWrite8,
+    /// transmit buffer empty
+    txempty: R::ReadOnly8,
+    /// receive buffer full
+    rxfull: R::ReadOnly8,
+}
+
+impl<R: LiteXSoCRegisterConfiguration> LiteXUartRegisters<R> {
+    /// Create an event manager instance for the UART events
+    fn ev<'a>(&'a self) -> LiteXUartEV<'a, R> {
+        LiteXUartEV::<R>::new(&self.ev_status, &self.ev_pending, &self.ev_enable)
+    }
+}
+
+register_bitfields![u8,
+    rxtx [
+        data OFFSET(0) NUMBITS(8) []
+    ],
+    txfull [
+        full OFFSET(0) NUMBITS(1) []
+    ],
+    rxempty [
+        empty OFFSET(0) NUMBITS(1) []
+    ],
+    txempty [
+        empty OFFSET(0) NUMBITS(1) []
+    ],
+    rxfull [
+        full OFFSET(0) NUMBITS(1) []
+    ]
+];
+
+pub struct LiteXUart<'a, R: LiteXSoCRegisterConfiguration> {
+    uart_regs: StaticRef<LiteXUartRegisters<R>>,
+    phy: Option<(StaticRef<LiteXUartPhyRegisters<R>>, u32)>,
+    tx_client: OptionalCell<&'a dyn uart::TransmitClient>,
+    rx_client: OptionalCell<&'a dyn uart::ReceiveClient>,
+    tx_buffer: TakeCell<'static, [u8]>,
+    tx_len: Cell<usize>,
+    tx_progress: Cell<usize>,
+    tx_aborted: Cell<bool>,
+    tx_deferred_call: Cell<bool>,
+    rx_buffer: TakeCell<'static, [u8]>,
+    rx_len: Cell<usize>,
+    rx_progress: Cell<usize>,
+    rx_aborted: Cell<bool>,
+    rx_deferred_call: Cell<bool>,
+    deferred_caller: &'static DynamicDeferredCall,
+    deferred_handle: OptionalCell<DeferredCallHandle>,
+}
+
+impl<'a, R: LiteXSoCRegisterConfiguration> LiteXUart<'a, R> {
+    pub const fn new(
+        uart_base: StaticRef<LiteXUartRegisters<R>>,
+        phy_args: Option<(StaticRef<LiteXUartPhyRegisters<R>>, u32)>,
+        deferred_caller: &'static DynamicDeferredCall,
+    ) -> LiteXUart<'a, R> {
+        LiteXUart {
+            uart_regs: uart_base,
+            phy: phy_args,
+            tx_client: OptionalCell::empty(),
+            rx_client: OptionalCell::empty(),
+            tx_buffer: TakeCell::empty(),
+            tx_len: Cell::new(0),
+            tx_progress: Cell::new(0),
+            tx_aborted: Cell::new(false),
+            tx_deferred_call: Cell::new(false),
+            rx_buffer: TakeCell::empty(),
+            rx_len: Cell::new(0),
+            rx_progress: Cell::new(0),
+            rx_aborted: Cell::new(false),
+            rx_deferred_call: Cell::new(false),
+            deferred_caller,
+            deferred_handle: OptionalCell::empty(),
+        }
+    }
+
+    pub fn initialize(&self, deferred_call_handle: DeferredCallHandle) {
+        self.uart_regs.ev().disable_all();
+        self.deferred_handle.set(deferred_call_handle);
+    }
+
+    pub fn transmit_sync(&self, bytes: &[u8]) {
+        // We need to make sure that we're not modifying interrupt
+        // pending and enabled bits here!
+        let regs = self.uart_regs;
```

Comment:
Don't save the registers in a temp value.

Reply:
Why shouldn't I do this? This reduces verbosity a bit and it only saves the reference to the register struct in a variable (so no real copying of registers is happening here).

Reply:
I think this is fine. The pattern we got rid of was `let regs = &*self.uart_regs`, which used to be necessary for some reason but no longer is.